<launch>
	<!-- FROBIT PARAMETERS --> 
	<param name="diff_steer_wheel_distance" value="0.375"/>
	<param name="ticks_per_meter_left" value="677" />
	<param name="ticks_per_meter_right" value="677" />

	<!-- FROBIT INTERFACE -->
	<group ns="fmControllers">
		<!--node pkg="sdu_frobit_v2" name="frobit" type="frobit_interface_node.py" output="screen" -->
		<node pkg="mr_frobit_interface" name="frobit" type="frobit_interface_node.py" output="screen" >
			<param name="actuation_enable_sub" value="/fmSafe/actuation_enable" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="/fmCommand/cmd_vel" value="True" />
			<param name="enc_left_pub" value="/fmInformation/encoder_left" />
			<param name="enc_right_pub" value="/fmInformation/encoder_right" />
			<param name="wheel_status_left_pub" value="/fmInformation/wheel_status_left" />
			<param name="wheel_status_right_pub" value="/fmInformation/wheel_status_right" />
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="wheel_pid_left_pub" value="/fmInformation/wheel_pid_left" />
			<param name="wheel_pid_right_pub" value="/fmInformation/wheel_pid_right" />
			<param name="nmea_from_frobit_sub" value="/fmSignal/nmea_from_frobit" />	 
			<param name="nmea_to_frobit_pub" value="/fmSignal/nmea_to_frobit" />	 
			<param name="publish_wheel_status_rate" value="10" />
			<param name="publish_wheel_feedback_rate" value="10" />
			<param name="publish_wheel_pid_rate" value="10" />
			<param name="castor_front" value="true"/>
		    <param name="max_linear_acceleration" value="16.2"/>
        	<param name="max_angular_acceleration" value="16.0"/>
			<param name="wheel_drive_feed_forward" value="85" /> 
			<param name="wheel_drive_kp" value="5" /> 
			<param name="wheel_drive_ki" value="10" /> 
			<param name="wheel_drive_kd" value="5" /> 
			<param name="wheel_drive_max_integral_output" value="300" /> 
			<param name="wheel_turn_feed_forward" value="110" /> 
			<param name="wheel_turn_kp" value="10" /> 
			<param name="wheel_turn_ki" value="20" /> 
			<param name="wheel_turn_kd" value="3" /> 
			<param name="wheel_turn_max_integral_output" value="300" /> 
			<param name="supply_voltage_scale_factor" value="0.03747" /> 
			<param name="min_supply_voltage" value="11.7" /> 
			<param name="show_voltage_interval" value="5" /> 
		</node>
	</group>

	<!-- SAFETY MODULE -->
	<group ns="fmSafety">
		<node pkg="mr_incident_handler" name="incident_handler" type="incidentHandlerNode.py" output="screen">
    		<param name="deadman_enable" value="True" />
    		<param name="deadman_timeout" value="0.100" />
    		<param name="deadman_sub" value="/fmSafe/deadman" />
    		<param name="critical_fault_enable" value="False" />
    		<param name="critical_fault_timeout" value="0.100" />
    		<param name="critical_fault_sub" value="/fmSafe/critical_fault" />
	    	<param name="obstacle_sub" value="/mrObstacleDetector/status" />
			<param name="actuation_enable_pub" value="/fmSafe/actuation_enable"/>
			<param name="nav_currentnode_sub" value="/mrNavigationController/currentNode"/>
		</node>
	</group>

	<!-- FROBIT SERIAL DEVICE AND PROTOCOL -->
	<node pkg="serial_nmea" name="frobit_nmea" type="serial_nmea_node" respawn="true" output="screen">
		<param name="nmea_from_device_pub" value="/fmSignal/nmea_from_frobit"/>
		<param name="nmea_to_device_sub" value="/fmSignal/nmea_to_frobit"/>
		<param name="serial_device" value="/dev/frobit"/>
		<param name="serial_baudrate" value="57600"/>
		<param name="use_nmea_checksum" value="true"/>
        </node>

	<!-- CALCULATE ODOMETRY (Based on wheel encoders and IMU yaw axis orientation) -->
	<group ns="fmProcessors">
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/encoder_left" />
			<param name="enc_right_sub" value="/fmInformation/encoder_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<param name="odom_pub" value="/fmKnowledge/pose" />
			<param name="odom_estimate_frame" value="/odom" />
			<param name="vehicle_frame" value="/base_link" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="/imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="z" />
		</node>
	</group>

	<!-- IMU SERIAL COMMUNICATION -->
	<node pkg="serial_string" name="imu_serial" type="serial_string_node" respawn="true" output="screen">
		<param name="publisher_topic" value="/fmData/imu_rx" />	 
		<param name="subscriber_topic" value="/fmData/imu_tx" />
		<param name="device" value="/dev/imu" />	
		<param name="baudrate" value="115200" />
		<param name="termination_character" value="10"/>
	</node> 

	<!-- Extract data from messages transmitted by the VectorNav VN-100 IMU -->
	<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen"  respawn="true">
		<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
		<param name="imu_pub" value="/fmInformation/imu"/>
		<param name="frame_id" value="imu_link"/>
		<param name="use_enu" value="true"/>
		<param name="covariance_x" value="0.001"/>
		<param name="covariance_y" value="0.001"/>
		<param name="covariance_z" value="0.001"/>
	 </node>

	<!-- MR Go -->
	<group ns="MR_Go">
    		<node pkg="mr_go" name="mr_go" type="mr_go" respawn="true" output="screen">
		    	<param name="odometry" value="/fmKnowledge/pose"/>
		    	<param name="pub_twist" value="/fmCommand/cmd_vel"/>
		    	<param name="pub_deadman" value="/fmSafe/deadman"/>
		    	<param name="srv_move" value="/mrGo/move"/>
		    	<param name="linear_precision" value="0.01"/>
		    	<param name="angular_precision" value="0.500"/>
		    	<param name="linear_speed" value="0.2"/>
		    	<param name="angular_speed" value="0.3"/>
    		 </node>
	</group>


</launch>

